File: src/main/java/com/vc/medium/L39.scala
package com.vc.medium

object L39 {
  def combinationSum(arr: Array[Int], target: Int): List[List[Int]] = {
    import scala.collection.mutable
    val res = new mutable.ListBuffer[List[Int]]()
    val list = new mutable.ArrayBuffer[Int]()
    val n = arr.length
    def solve(start: Int, remain: Int): Unit = {
      if(remain == 0) res += list.toList
      else if(remain > 0){
        (start until n).foreach(i => {
          list += arr(i)
          solve(i, remain - arr(i))
          list.remove(list.length - 1)
        })
      }
    }
    solve(0, target)
    res.toList
  }
}

File: src/main/java/com/vc/medium/L39.java
package com.vc.medium;

import java.util.ArrayList;
import java.util.List;

class L39 {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        this.res = new ArrayList<>();
        if(candidates.length == 0) return res;
        helper(candidates, 0, target, new ArrayList<>());
        return res;
    }

    private void helper(int[] candidates, int index, int target, List<Integer> list) {
        if(target < 0) return;
        if(target == 0) res.add(new ArrayList<>(list));

        for(int i = index; i < candidates.length; i++) {
            list.add(candidates[i]);
            helper(candidates, i, target - candidates[i], list);
            list.remove(list.size() - 1);
        }
    }
}
